### PR TITLE
US #13.5 - Dynamic of range actions available in several instants

### PR DESCRIPTION
**Pull request by phiedw**
**Please check if the PR fulfills these requirements** *(please use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes)*
- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes / features)

**Does this PR already have an issue describing the problem ?** *If so, link to this issue using `'#XXX'` and skip the rest*
No

**What kind of change does this PR introduce?** *(Bug fix, feature, docs update, ...)*
feature / bug fix

**What is the current behavior?** *(You can also link to an open issue here)*
types of range are not clearly defined
also issue where the dynamic ranges can update their limits while a perimeter is being optimized

**What is the new behavior (if this is a feature change)?**
type of ranges are more clearly defined, clean crac makes some extra checks
also fixed the updating issue

**Does this PR introduce a breaking change?** *(What changes might users need to make in their application due to this PR?)*

**Other information**:

(if any of the questions/checkboxes don't apply, please delete them entirely)

### CODE REVIEW

File: data/crac/crac-api/src/main/java/com/farao_community/farao/data/crac_api/Range.java [hunk 1]
```diff
@@ -0,0 +1,11 @@
+package com.farao_community.farao.data.crac_api;
+
+public interface Range {
```
Comment (baptsegu) [inline data/crac/crac-api/src/main/java/com/farao_community/farao/data/crac_api/Range.java:3]: What do you think about putting also PstRange and RangeDefinition in crac-api ? (see previous comment)

File: data/crac/crac-api/src/main/java/com/farao_community/farao/data/crac_api/Range.java [hunk 2]
```diff
@@ -0,0 +1,11 @@
+package com.farao_community.farao.data.crac_api;
+
```
Comment (baptsegu) [inline data/crac/crac-api/src/main/java/com/farao_community/farao/data/crac_api/Range.java:2]: missing headers

File: data/crac/crac-util/src/main/java/com/farao_community/farao/data/crac_util/CracCleaner.java [hunk 3]
```diff
@@ -142,6 +143,24 @@ public CracCleaner() {
         removedStates.forEach(state -> crac.removeState(state.getId()));
         report.forEach(LOGGER::warn);
 
+        // remove PstRanges with a RELATIVE_TO_PREVIOUS_INSTANT for preventive PST RAs
```
Comment (baptsegu) [inline data/crac/crac-util/src/main/java/com/farao_community/farao/data/crac_util/CracCleaner.java:146]: maybe add a unit test on this behaviour

File: data/crac/crac-impl/src/main/java/com/farao_community/farao/data/crac_impl/remedial_action/range_action/PstWithRange.java [hunk 4]
```diff
@@ -28,7 +28,7 @@
  * @author Viktor Terrier {@literal <viktor.terrier at rte-france.com>}
  */
 @JsonTypeName("pst-with-range")
-public final class PstWithRange extends AbstractRangeAction implements PstRange {
+public final class PstWithRange extends AbstractRangeAction implements PstRangeAction {
```
[the remark this answers is not in the document]
Comment (phiedw) [reply data/crac/crac-impl/src/main/java/com/farao_community/farao/data/crac_impl/remedial_action/range_action/PstWithRange.java:31]: indeed, i never liked that name either :stuck_out_tongue:

File: data/crac/crac-util/src/main/java/com/farao_community/farao/data/crac_util/CracCleaner.java (hunk 3 above)
Comment (phiedw) [reply data/crac/crac-util/src/main/java/com/farao_community/farao/data/crac_util/CracCleaner.java:146]: added a test for this in the testCleanCrac unit test

File: data/crac/crac-api/src/main/java/com/farao_community/farao/data/crac_api/Range.java (hunk 2 above)
Comment (phiedw) [reply data/crac/crac-api/src/main/java/com/farao_community/farao/data/crac_api/Range.java:2]: fixed

File: data/crac/crac-api/src/main/java/com/farao_community/farao/data/crac_api/Range.java (hunk 1 above)
Comment (phiedw) [reply data/crac/crac-api/src/main/java/com/farao_community/farao/data/crac_api/Range.java:3]: seems like a good idea, RangeDefinition is already in the API so PstRange has good reasons to be too

**Review by baptsegu — APPROVED** [review]
I'm still up to moving PstRange to the crac-api, even if we cannot do all the adders within this PR.

But if you think it's not pertinent, I'm ok with all the rest !

File: ra-optimisation/rao-commons/src/main/java/com/farao_community/farao/rao_commons/CracResultManager.java [hunk 5]
```diff
@@ -100,10 +98,13 @@ public void fillRangeActionResultsWithLinearProblem(LinearProblem linearProblem)
...
                 int approximatedPostOptimTap;
                 double approximatedPostOptimAngle;
-                if (raoData.getAvailableRangeActions().contains(rangeAction) && linearProblem.getRangeActionSetPointVariable(rangeAction) != null) {
-                    String networkElementId = rangeAction.getNetworkElements().iterator().next().getId();
-                    double rangeActionVal = linearProblem.getRangeActionSetPointVariable(rangeAction).solutionValue();
-                    PstRangeAction pstRangeAction = (PstRangeAction) rangeAction;
+                String networkElementId = rangeAction.getNetworkElements().iterator().next().getId();
+                Optional<RangeAction> sameNetworkElementRangeAction = raoData.getAvailableRangeActions().stream()
+                    .filter(ra -> ra.getNetworkElements().iterator().next().getId().equals(networkElementId) && linearProblem.getRangeActionSetPointVariable(ra) != null)
+                    .findFirst();
```
Comment (baptsegu) [inline ra-optimisation/rao-commons/src/main/java/com/farao_community/farao/rao_commons/CracResultManager.java:104]: Seb usually advise to use `findAny()` in that case, which is faster

**Review by baptsegu — APPROVED** [review]

**Review by baptsegu — APPROVED** [review]

**Pull request merged**